src/git.js

```diff
@@ -243,56 +235,70 @@ export default class Git {
-	async getTree(commitSha) {
-		const output = await execCmd(
-			`git ls-tree -r --full-tree ${ commitSha }`,
+	async getTreeId(commitSha) {
+		core.debug(`Getting treeId for commit ${ commitSha }`)
+		const output = (await execCmd(
+			`git cat-file -p ${ commitSha }`,
 			this.workingDir
-		)
+		)).split('\n')
 
-		const tree = []
-		for (const treeObject of output.split('\n')) {
-			const [ mode, type, sha ] = treeObject.split(/\s/)
-			const file = treeObject.split('\t')[1]
-
-			const treeEntry = {
-				mode,
-				type,
-				sha,
-				path: file
-			}
-
-			tree.push(treeEntry)
-		}
+		const commitHeaders = output.slice(0, output.findIndex((e) => e === ''))
+		const tree = commitHeaders.find((e) => e.startsWith('tree')).replace('tree ', '')
 
 		return tree
 	}
 
-	// Creates the blob objects in GitHub for the files that are not in the previous commit only
-	async createGithubBlobs(commitSha) {
-		core.debug('Creating missing blobs on GitHub')
-		const [ previousTree, tree ] = await Promise.all([ this.getTree(`${ commitSha }~1`), this.getTree(commitSha) ])
-		const promisesGithubCreateBlobs = []
-
-		for (const treeEntry of tree) {
-			// If the current treeEntry are in the previous tree, that means that the blob is uploaded and it doesn't need to be uploaded to GitHub again.
-			if (previousTree.findIndex((entry) => entry.sha === treeEntry.sha) !== -1) {
-				continue
-			}
-
-			const base64Content = await this.getBlobBase64Content(treeEntry.path)
+	async getTreeDiff(referenceTreeId, differenceTreeId) {
+		const output = await execCmd(
+			`git diff-tree ${ referenceTreeId } ${ differenceTreeId } -r`,
+			this.workingDir
+		)
 
-			// Creates the blob. We don't need to store the response because the local sha is the same and we can use it to reference the blob
-			const githubCreateBlobRequest = this.github.git.createBlob({
-				owner: this.repo.user,
-				repo: this.repo.name,
-				content: base64Content,
-				encoding: 'base64'
+		const diff = []
+		for (const line of output.split('\n')) {
+			const splitted = line
+				.replace(/^:/, '')
+				.replace('\t', ' ')
+				.split(' ')
+
+			const [
+				newMode,
+				previousMode,
+				newBlob,
+				previousBlob,
+				change,
+				path
+			] = splitted
+
+			diff.push({
+				newMode,
+				previousMode,
+				newBlob,
+				previousBlob,
+				change,
+				path
 			})
-			promisesGithubCreateBlobs.push(githubCreateBlobRequest)
 		}
 
-		// Wait for all the file uploads to be completed
-		await Promise.all(promisesGithubCreateBlobs)
+		return diff
+	}
+
+	// Creates the blob objects in GitHub for the files that are not in the previous commit only
+	async uploadGitHubBlob(blob) {
+		core.debug(`Uploading GitHub Blob for blob ${ blob }`)
+		const fileContent = await execCmd(
+			`git cat-file -p ${ blob }`,
+			this.workingDir,
+			false
+		)
+
+		// Creates the blob. We don't need to store the response because the local sha is the same and we can use it to reference the blob
```

Comment:
Hi,
We have integrated this PR as it solves a couple of issues for us, however - while I can't find documentation on it - we have observed a  behavior (that happens rarely) where the local sha is different from the sha returned by github api. I know it makes no sense given git internals, not sure what exactly happens here, probably some transformation on the content/filename by the API.

So I would amend the code to use the remote sha

Reply:
Hi @sthuck.

When did you find the problem? I found the same problem that you described after publishing the pull request,  but I solved it (or I thought I did at least) the day 11/09. I forced pushed the commit to my branch replacing the previous commit.

The problem was caused by my usage of the ExecCmd function which by default trims the output of the command. In this case, I use it to get the content of a blob in the local git,  which is a problem for the files is a problem for files which end with empty lines. I solved it passing the parameter to disable the trimming of the command (https://github.com/alvarezfr/repo-file-sync-action/commit/0fad1b17442ae07ba5fcc763de16d367fb6a51f9#diff-30246b15f18bb39311579208583260447efd5ee017c25172393ad4505bb8c5a4R292). Based on my tests this solves the issue.